src/library/rekordbox/rekordboxfeature.cpp

```diff
@@ -0,0 +1,971 @@
+    QString artist = artistsMap[track->artist_id()];
+    QString album = albumsMap[track->album_id()];
+    QString year = QString::number(track->year());
+    QString genre = genresMap[track->genre_id()];
+    QString location = devicePath + getText(track->file_path());
+    float bpm = static_cast<float>(track->tempo() / 100.0);
+    int bitrate = static_cast<int>(track->bitrate());
+    QString key = keysMap[track->key_id()];
+    int playtime = static_cast<int>(track->duration());
+    int rating = static_cast<int>(track->rating());
+    QString comment = getText(track->comment());
+    QString tracknumber = QString::number(track->track_number());
+    QString anlzPath = devicePath + getText(track->analyze_path());
+
+    query.bindValue(":rb_id", rbID);
+    query.bindValue(":artist", artist);
+    query.bindValue(":title", title);
+    query.bindValue(":album", album);
+    query.bindValue(":genre", genre);
+    query.bindValue(":year", year);
+    query.bindValue(":duration", playtime);
+    query.bindValue(":location", location);
+    query.bindValue(":rating", rating);
+    query.bindValue(":comment", comment);
+    query.bindValue(":tracknumber", tracknumber);
+    query.bindValue(":key", key);
+    query.bindValue(":bpm", bpm);
+    query.bindValue(":bitrate", bitrate);
+    query.bindValue(":analyze_path", anlzPath);
+    query.bindValue(":device", device);
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+    }
+
+    int trackID = -1;
+    QSqlQuery finderQuery(database);
+    finderQuery.prepare("select id from rekordbox_library where rb_id=:rb_id and device=:device");
+    finderQuery.bindValue(":rb_id", rbID);
+    finderQuery.bindValue(":device", device);
+
+    if (!finderQuery.exec()) {
+        LOG_FAILED_QUERY(finderQuery)
+                << "rbID:" << rbID;
+    }
+
+    if (finderQuery.next()) {
+        trackID = finderQuery.value(finderQuery.record().indexOf("id")).toInt();
+    }
+
+    // Insert into device all tracks playlist
+    queryInsertIntoDevicePlaylistTracks.bindValue(":track_id", trackID);
+    queryInsertIntoDevicePlaylistTracks.bindValue(":position", audioFilesCount);
+
+    if (!queryInsertIntoDevicePlaylistTracks.exec()) {
+        LOG_FAILED_QUERY(queryInsertIntoDevicePlaylistTracks)
+                << "trackID:" << trackID
+                << "position:" << audioFilesCount;
+    }
+}
+
+void buildPlaylistTree(
+        QSqlDatabase &database,
+        TreeItem *parent,
+        uint32_t parentID,
+        QMap<uint32_t, QString> &playlistNameMap,
+        QMap<uint32_t, bool> &playlistIsFolderMap,
+        QMap<uint32_t, QMap<uint32_t, uint32_t>> &playlistTreeMap,
+        QMap<uint32_t, QMap<uint32_t, uint32_t>> &playlistTrackMap,
+        QString playlistPath,
+        QString device);
+
+QString parseDeviceDB(TrackCollection *trackCollection, TreeItem *deviceItem) {
+    QSqlDatabase database = QSqlDatabase::cloneDatabase(trackCollection->database(),
```

Comment:
Ok, cloning the database explicitly should work.

A safer alternative would be to use `Library::dbConnectionPool()` to obtain a thread-local database connection directly in the worker thread and only when needed.

Reply:
Instead of passing `TrackCollection*` that is not thread-safe and strictly bound to the main thread this function should only receive a thread-safe `mixxx::DbConnectionPoolPtr` for accessing the database:

```
QSqlDatabase database = mixxx::DbConnectionPooled(pDbConnectionPool);
```

Reply:
The *pooler* was necessary, because we cannot add any hooks into the internal lifetime management of QSqlDatabase.